File: core/src/nl/t64/game/rpg/screens/world/FogOfWar.java
package nl.t64.game.rpg.screens.world;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.math.Circle;
import com.badlogic.gdx.math.Vector2;
import lombok.NoArgsConstructor;
import nl.t64.game.rpg.constants.Constant;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;


class FogOfWar {

    private static final float FOG_OF_WAR_RADIUS = (Constant.TILE_SIZE * 6) - 1;

    private final Map<String, Set<FogPoint>> container = new HashMap<>();

    void putIfAbsent(GameMap gameMap) {
        container.putIfAbsent(gameMap.mapTitle, fillFogOfWar(gameMap));
    }

    void update(Vector2 playerPosition, String mapTitle) {
        var sightRadius = new Circle(playerPosition, FOG_OF_WAR_RADIUS);
        container.get(mapTitle)
                 .stream()
                 .filter(sightRadius::contains)
                 .forEach(FogPoint::setExplored);
    }

    void draw(ShapeRenderer shapeRenderer, String mapTitle) {
        shapeRenderer.setColor(Color.BLACK);
        container.get(mapTitle)
                 .stream()
                 .filter(FogPoint::isUnexplored)
                 .forEach(point -> shapeRenderer.rect(point.x, point.y,
                                                      Constant.HALF_TILE_SIZE, Constant.HALF_TILE_SIZE));
    }

    private Set<FogPoint> fillFogOfWar(GameMap gameMap) {
        return IntStream.range(0, gameMap.getWidth())
                        .mapToObj(x -> getFogOfWarStream(x, gameMap))
                        .flatMap(Function.identity())
                        .collect(Collectors.toSet());
    }

    private Stream<FogPoint> getFogOfWarStream(int x, GameMap gameMap) {
        return IntStream.range(0, gameMap.getHeight())
                        .mapToObj(y -> new FogPoint(x, y));
    }

    @NoArgsConstructor
    private static class FogPoint extends Vector2 {

        private boolean isExplored;

        private FogPoint(float x, float y) {
            super(x * Constant.HALF_TILE_SIZE, y * Constant.HALF_TILE_SIZE);
            this.isExplored = false;
        }

        private void setExplored() {
            isExplored = true;
        }

        private boolean isUnexplored() {
            return !isExplored;
        }

    }

}

File: core/src/nl/t64/game/rpg/screens/world/FogOfWar.kt
package nl.t64.game.rpg.screens.world

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.badlogic.gdx.math.Circle
import com.badlogic.gdx.math.Vector2
import nl.t64.game.rpg.constants.Constant


private const val FOG_OF_WAR_RADIUS = Constant.TILE_SIZE * 6 - 1

internal class FogOfWar {

    private val container: MutableMap<String, Set<FogPoint>> = HashMap()

    fun putIfAbsent(gameMap: GameMap) {
        container.putIfAbsent(gameMap.mapTitle, fillFogOfWar(gameMap))
    }

    fun update(playerPosition: Vector2, mapTitle: String) {
        val sightRadius = Circle(playerPosition, FOG_OF_WAR_RADIUS)
        container[mapTitle]!!
            .filter { sightRadius.contains(it) }
            .forEach { it.isExplored = true }
    }

    fun draw(shapeRenderer: ShapeRenderer, mapTitle: String) {
        shapeRenderer.color = Color.BLACK
        container[mapTitle]!!
            .filter { !it.isExplored }
            .forEach { shapeRenderer.rect(it.x, it.y, Constant.HALF_TILE_SIZE, Constant.HALF_TILE_SIZE) }
    }

    private fun fillFogOfWar(gameMap: GameMap): Set<FogPoint> {
        return (0 until gameMap.width)
            .map { createFogPoints(it, gameMap) }
            .flatten()
            .toSet()
    }

    private fun createFogPoints(x: Int, gameMap: GameMap): Set<FogPoint> {
        return (0 until gameMap.height).map { FogPoint(x.toFloat(), it.toFloat()) }.toSet()
    }

}

private class FogPoint(
    x: Float = 0f, y: Float = 0f
) : Vector2(x * Constant.HALF_TILE_SIZE, y * Constant.HALF_TILE_SIZE) {
    var isExplored = false
}
